### PR TITLE
[extractor/common] Fix `json_ld` type checks

### LINKED ISSUE #5144: `InfoExtractor._json_ld` fails under certain conditions
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.10.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

`InfoExtractor._json_ld` fails under certain conditions.
Inside `traverse_json_ld` `e` is assumed a dict eventhough it does not always have to be.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://www.autoweek.nl/autotests/artikel/porsche-911-gt3-rs-rij-impressie-2/']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.10.04 [4e0511f27] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Python 3.10.5 (CPython 64bit) - Windows-10-10.0.22000-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg n5.1-1-g6d83c5f876-20220727 (setts), ffprobe n5.1-1-g6d83c5f876-20220727
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2022.09.14, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] Loaded 1691 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.10.04, Current version: 2022.10.04
yt-dlp is up to date (2022.10.04)
[debug] [generic] Extracting URL: https://www.autoweek.nl/autotests/artikel/porsche-911-gt3-rs-rij-impressie-2/
[generic] porsche-911-gt3-rs-rij-impressie-2: Downloading webpage
[redirect] Following redirect to https://myprivacy.dpgmedia.nl/consent?siteKey=ujm6mv0jrqiz5syr&callbackUrl=https%3A%2F%2Fwww.autoweek.nl%2Fprivacygate-confirm%3FredirectUri%3D%252Fautotests%252Fartikel%252Fporsche-911-gt3-rs-rij-impressie-2%252F
[debug] [generic] Extracting URL: https://myprivacy.dpgmedia.nl/consent?siteKey=ujm6mv0jrqiz5syr&callbackUrl=https%3A%2F%2Fwww.autoweek.nl%2Fprivacygate-confirm%3FredirectUri%3D%252Fautotests%252Fartikel%252Fporsche-911-gt3-rs-rij-impressie-2%252F
[generic] consent?siteKey=ujm6mv0jrqiz5syr&callbackUrl=https://www.autoweek: Downloading webpage
[redirect] Following redirect to https://www.autoweek.nl/autotests/artikel/porsche-911-gt3-rs-rij-impressie-2/
[debug] [generic] Extracting URL: https://www.autoweek.nl/autotests/artikel/porsche-911-gt3-rs-rij-impressie-2/
[generic] porsche-911-gt3-rs-rij-impressie-2: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] porsche-911-gt3-rs-rij-impressie-2: Extracting information
[debug] Looking for Brightcove embeds
[debug] Looking for embeds
ERROR: 'list' object has no attribute 'get'
Traceback (most recent call last):
  File "D:\Documents\GitHub\Grub4K\yt-dlp\yt_dlp\YoutubeDL.py", line 1477, in wrapper
    return func(self, *args, **kwargs)
  File "D:\Documents\GitHub\Grub4K\yt-dlp\yt_dlp\YoutubeDL.py", line 1553, in __extract_info
    ie_result = ie.extract(url)
  File "D:\Documents\GitHub\Grub4K\yt-dlp\yt_dlp\extractor\common.py", line 672, in extract
    ie_result = self._real_extract(url)
  File "D:\Documents\GitHub\Grub4K\yt-dlp\yt_dlp\extractor\generic.py", line 2897, in _real_extract
    json_ld = self._search_json_ld(webpage, video_id, default={})
  File "D:\Documents\GitHub\Grub4K\yt-dlp\yt_dlp\extractor\common.py", line 1450, in _search_json_ld
    info = self._json_ld(
  File "D:\Documents\GitHub\Grub4K\yt-dlp\yt_dlp\extractor\common.py", line 1631, in _json_ld
    traverse_json_ld(json_ld)
  File "D:\Documents\GitHub\Grub4K\yt-dlp\yt_dlp\extractor\common.py", line 1576, in traverse_json_ld
    traverse_json_ld(variadic(e['@graph'], allowed_types=(dict,)), at_top_level=False)
  File "D:\Documents\GitHub\Grub4K\yt-dlp\yt_dlp\extractor\common.py", line 1624, in traverse_json_ld
    video = e.get('video')
AttributeError: 'list' object has no attribute 'get'
```

### PR DESCRIPTION
**Pull request by Grub4K**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

Secures `traverse_json_ld` against unexpected types.

Fixes #5144, #5143

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by Grub4K** [comment]
The `json_ld` provided has a graph looking like this: `[{"@graph": [{...}, [{...}]]}]`.
Currently the nested list inside the `@graph` list doesn't get traversed.
I am not sure if it makes make sense to allow such nesting. I have a patch already written, but on the other hand, traversing deeper might give unrelated things we wouldn't want.

**Comment by pukkandan** [comment]
> I am not sure if it makes make sense to allow such nesting

Does it make sense for the example you are looking at? Or, does it end up giving redundant/useless info? We should implement this only if we have actual examples where this is useful

**Comment by Grub4K** [comment]
It does not change the result for this case, I just thought it might make sense for potential future cases.

If you think it is better to wait for a specific case then this is good to go.

File: yt_dlp/extractor/common.py [hunk 1]
```diff
@@ -1569,11 +1565,18 @@ def extract_video_object(e):
             extract_chapter_information(e)
 
         def traverse_json_ld(json_ld, at_top_level=True):
+            if isinstance(json_ld, dict):
+                json_ld = [json_ld]
+            if not isinstance(json_ld, list):
+                return
+
             for e in json_ld:
```
Comment (pukkandan) [inline yt_dlp/extractor/common.py:1573]:
```suggestion
            for e in variadic(json_ld):
```
accomplishes the same

File: yt_dlp/extractor/common.py [hunk 2]
```diff
@@ -1463,13 +1463,9 @@ def _search_json_ld(self, html, video_id, expected_type=None, *, fatal=True, def
     def _json_ld(self, json_ld, video_id, fatal=True, expected_type=None):
         if isinstance(json_ld, str):
             json_ld = self._parse_json(json_ld, video_id, fatal=fatal)
-        if not json_ld:
+        if not json_ld or not isinstance(json_ld, (list, dict)):
             return {}
```
Comment (pukkandan) [inline yt_dlp/extractor/common.py:1467]: this also seems unnecesary. The check in L1574 should handle all of these

**Pull request merged**